### ISSUE TITLE
Errors when running with Babel-based React Compiler

### ISSUE THREAD

**Issue by jgarplind** [remix-run/react-router#12352]
### Describe what's incorrect/missing in the documentation

https://reactrouter.com/upgrading/router-provider#2-install-the-vite-plugin describes how to adopt the new React Router framework, coming from RouterProvider.

I started this migration, but ran into issues when I realized that the `reactRouter` plugin does not accept the same options as the `react` plugin, which stopped me in my tracks.

Specifically, I have the need to add babel plugins both for React compiler: https://react.dev/learn/react-compiler#usage-with-vite and Lingui: https://lingui.dev/tutorials/setup-vite.

In Discord, Kosire suggested I could patch the plugin: https://discord.com/channels/770287896669978684/1309472866453946368/1309472866453946368, which would probably work, but is not very accessible to the average user of the library, and would also be a significant maintenance burden over time, so it's not something I would like to do without being sure it is the only way.

Is this the state of things? Or is there a more appropriate way to handle this scenario, which could then be added to the documentation?

**Comment by timdorr** [comment]
Currently, [it takes no options](https://github.com/remix-run/react-router/blob/ae79fcec6646a127fed133937f246ce5343bd641/packages/react-router-dev/vite/plugin.ts#L403), so that would be the place to start. Unfortunately, it doesn't look like an easy task.

**Comment by jgarplind** [comment]
Thanks for weighing in! Yeah, that much is clear from looking at the source code. Whether it is an easy task or not is harder for me to tell.

I'm not sure why/if it would be difficult to accept the same arguments as the `react`-plugin (or at least the `babel` option). I wrote this issue hoping it was more or less forgotten, but of course it seems more likely that there are legitimate technical reasons.

**Comment by romulovalez** [comment]
I also need to pass the babel option in order to use styled-components, so I think is a must for a lot of users

**Comment by faradaytrs** [comment]
We also need babel options because we are using react compiler

**Comment by MylesWardell** [comment]
Bump Babbel for react-intl

**Comment by AlemTuzlak** [comment]
I'll shortly send a PR link on how to add this to your project

**Comment by AlemTuzlak** [comment]
https://github.com/forge42dev/supabase-realtime/pull/6
Here is a reference implementation that successfully adds it into a react router v7 framework mode project. also attached are the images of all the relevant changes 
![package json](https://github.com/user-attachments/assets/7764787d-603e-4a0d-8c62-c1eb832ebff8)
![vite config ts](https://github.com/user-attachments/assets/073684e2-692e-445b-9e62-3cfffe6b8678)
![babelrc](https://github.com/user-attachments/assets/b00774af-7776-4dbe-bd09-ffe9115cbf1f)

**Comment by jgarplind** [comment]
Thanks @AlemTuzlak, looks promising! I'll report back here once I've had the opportunity to test it, and confirmed that it covers my use cases, including Lingui.

**Comment by jgarplind** [comment]
Documenting as I go:
- Install `vite-plugin-babel` as per instructions above
- Install plugins as desired (see https://lingui.dev/releases/migration-5#standalone-babel-plugin-lingui-macro for Lingui)
- Reference them directly in your Vite config like so:
```
export default defineConfig({
  plugins: [
    babel({
      babelConfig: {
        plugins: ["@lingui/babel-plugin-lingui-macro"],
      },
    }),
    ...
```

This unblocks me and lets me continue my migration (which I don't have the time to finalize right now).

Unclear so far for me if I need to add `@babel/preset-typescript`.

I'll update this later as I find the time to finalize the migration, but this is very promising.

Hopefully we can add this to the documentation once we've validated it works widely.

**Comment by timdorr** [comment]
Note that having multiple plugins will instantiate multiple copies of Babel, transform your code multiple times, and result in both memory bloat and slower performance. That is why the official React plugin takes these options. We should do the same so builds remain as fast as possible.

**Comment by AlemTuzlak** [comment]
@timdorr I'm not sure that's the case? It runs a single transform with multiple plugins from what I saw while trying it out?

**Comment by timdorr** [comment]
The React Router plugin runs its own Babel instance and transforms, depending on the module type:

https://github.com/remix-run/react-router/blob/9819f96c85aa9bb8ba2dc660c0ef9baa0e32c048/packages/react-router-dev/vite/plugin.ts#L1444-L1453

https://github.com/remix-run/react-router/blob/9819f96c85aa9bb8ba2dc660c0ef9baa0e32c048/packages/react-router-dev/vite/plugin.ts#L1523-L1534

But this is also what the official React plugin does:

https://github.com/vitejs/vite-plugin-react/blob/2e368a65aaf8314142538c369c296e277fdfb69b/packages/plugin-react/src/index.ts#L232-L258

So, you're creating at least two instances of Babel. It would be far better for us to just take options directly so that transform step is done one time.

**Comment by GrahamQuan** [comment]
This would be useful for react compiler

**Comment by crispiestsquid** [comment]
Stumbled upon this while also hoping for a simple way to implement React Compiler.

Would love to see the `reactRouter` plugin take options. If there is anything I can do to help move that along please reach out.

Otherwise I will keep playing with the workaround while patiently wait for this

**Comment by sindras** [comment]
Thanks for starting this thread. I also noticed that when I'm migrating from Remix to RRv7, there is no option to pass presets, which makes it difficult to implement PWA using [@vite-pwa/remix](https://vite-pwa-org.netlify.app/frameworks/remix.html).

**Comment by kazem3d** [comment]
> Thanks for starting this thread. I also noticed that when I'm migrating from Remix to RRv7, there is no option to pass presets, which makes it difficult to implement PWA using [@vite-pwa/remix](https://vite-pwa-org.netlify.app/frameworks/remix.html).

I also have this problem,How to have a pwa application in RRv7 ?

**Comment by sdavids** [comment]
Stumbled upon this issue via:

https://react.dev/learn/react-compiler#usage-with-vite
https://react.dev/learn/react-compiler#usage-with-remix

**Comment by mikkpokk** [comment]
> Stumbled upon this issue via:
> 
> https://react.dev/learn/react-compiler#usage-with-vite https://react.dev/learn/react-compiler#usage-with-remix

Doesn't work with RR7 as expected. Causes randomly _Invalid Hook Call_ errors during navigation.

**Comment by WilliamPeralta** [comment]
nothing yet?

**Comment by jgarplind** [comment]
Personally I'm hoping that "[just a Vite plugin again](https://bsky.app/profile/jgarplind.bsky.social/post/3lhnty4pc5s2e)" might enable this use case again.

**Comment by mikkpokk** [comment]
> Personally I'm hoping that "[just a Vite plugin again](https://bsky.app/profile/jgarplind.bsky.social/post/3lhnty4pc5s2e)" might enable this use case again.

The Vite Experimental Environment API is not related at all with the React Compiler.

[Add unstable Vite Environment API support #12936](https://github.com/remix-run/react-router/pull/12936)

[Vite 6 Experimental Environment API](https://vite.dev/guide/api-environment)

**Comment by jgarplind** [comment]
> The Vite Experimental Environment API is not related at all with the React Compiler or any other optional options of React.

I don't want to start a long discussion here, and I agree they are not directly related.

My thinking is that this limitation was introduced when (and because) React Router went _away_ from being "just a Vite plugin". If we go back to "just a Vite plugin", it seems reasonable that we can freely include Vite plugins of our choice again.

This has yet to be confirmed by anyone.

**Comment by mikkpokk** [comment]
It's ok, that's why I answered to clarify things out.

Before #12936, configuring CSR, SSR, or any other custom build environment in RR7 required using the `react-router` CLI instead of the `vite:build` CLI command.

With #12936, this is now possible using `vite:build`, including support for Vite's experimental environment API. 

**However, it does not affect React or React Router plugin options.**

<img width="883" alt="Image" src="https://github.com/user-attachments/assets/b95e3259-f84b-4718-b192-e3ec10158a73" />

**Comment by madisvain** [comment]
@jgarplind did you manage to get linguijs working with RR7? Even with vite-plugin-babel I did not manage to get this to work sadly.

**Comment by pawelblaszczyk5** [comment]
> [@jgarplind](https://github.com/jgarplind) did you manage to get linguijs working with RR7? Even with vite-plugin-babel I did not manage to get this to work sadly.
> 
> `[vite] Named export 'useLingui' not found. The requested module '@lingui/react/macro' is a CommonJS module, which may not support all module.exports as named exports. CommonJS modules can always be imported via the default export, for example using:`

I’ve made it working without issues - here’s my bite config:

```ts

import { lingui } from "@lingui/vite-plugin";
import { reactRouter } from "@react-router/dev/vite";
import { reactRouterHonoServer } from "react-router-hono-server/dev";
import { defineConfig } from "vite";
import babel from "vite-plugin-babel";

const babelConfig = {
	plugins: [
		["@lingui/babel-plugin-lingui-macro", {}],
		["babel-plugin-react-compiler", {}],
		["@babel/plugin-syntax-jsx", {}],
	],
	presets: ["@babel/preset-typescript"],
};

export default defineConfig({
	plugins: [
		reactRouterHonoServer(),
		reactRouter(),
		babel({
			babelConfig,
			filter: /\.[jt]sx?$/u,
			loader: "jsx",
		}),
		lingui(),
	],
});

```

**Comment by pawelblaszczyk5** [comment]
> [@madisvain](https://github.com/madisvain) I'm using a similar config as [@pawelblaszczyk5](https://github.com/pawelblaszczyk5), which in my case (and supposedly also in their case) means that while we're using RR7, we are not able to use framework mode (yet).

I’m using framework mode 😃

**Comment by orzechowskid** [comment]
@pedsmoreira 's solution described [above](https://github.com/remix-run/react-router/issues/12352#issuecomment-2701558448) works great for me with `react@^19.0.0`, `react-router@^7.2.0`, and `styled-components@^6.1.15`.

**Comment by mikkpokk** [comment]
> and on `entry.server.tsx` - make sure to use `onAllReady` instead of `onShellReady`

It's important to mention, you disable streaming abilities by replacing `onShellReady` with `onAllReady`.

**Comment by pedsmoreira** [comment]
@mikkpokk that's an important point indeed

In my case, I'm trying to port a large application from a totally custom setup to react-router v7 framework. The previous setup used `renderToString` so getting to use the new streaming APIs, although not leveraging steaming fully, is already a great step ahead.

I imagine that there are other folks in a similar situation,.

I'm hoping that eventually someone will figure out and share how to integrate styled-components properly, or the library's authors will release an update for it

**Comment by pcattori** [comment]
Babel is an implementation detail of how HMR/React Refresh works in React Router that we may change, so adding options to integrate with our internal usage of Babel is not something we plan to do.

To apply additional transforms, you can use Vite plugins. For example, to use the React Compiler, you can use `vite-plugin-babel`. It is true that this means that _two_ separate Babel instances may be present but so far this doesn't seem to be a cause for concern. In the React Compiler use-case, you'd want to run the compiler in development as well as in production, but our internal Babel instance only runs in development (since its only used for HMR/React Refresh). With Vite's module graph ensuring that transforms are cached and only run as-needed in development, there hasn't been any meaningful perf degradations from having two Babel instances. Though if you do run into such a use-case, we'd love to hear about it (in a separate issue please)!

So in short, React Router composes well with existing Vite plugins so if you need to run Babel transforms, you can do so via `vite-plugin-babel`. If there are any difficulties with this, we can definitely open up discussions and add docs, but so far none of the difficulties mentioned seem to be React Router bugs, so I'm closing this issue.

**Comment by mikkpokk** [comment]
> > Stumbled upon this issue via:
> > https://react.dev/learn/react-compiler#usage-with-vite https://react.dev/learn/react-compiler#usage-with-remix
> 
> Doesn't work with RR7 as expected. Causes randomly _Invalid Hook Call_ errors during navigation.

@pcattori I don't think such errors are intended on the RR7 side.

The error is probably caused because the second Babel plugin overrides the RR7-compiled code, but it isn't cached to the `.react-router` directory as expected. This leads to _Invalid Hook Call_ errors. A forced reload, of course, clears the error, but we can't really consider that an expected or acceptable failure.

Normally, the `unstable_optimizeDeps` option would fix such an issue, but not in this case. With the second Babel plugin and `react-compiler` enabled, `unstable_optimizeDeps` ends up breaking the project render entirely.

**Comment by pcattori** [comment]
@mikkpokk yea that sounds like a bug, but different from the claim that "React Router plugin should accept Babel options". Mind opening up a new issue for that?

**Comment by mikkpokk** [comment]
Isn't it more practical to rename the issue title, if that's the case?

There are six months' worth of discussion and proposed workarounds (which don't entirely work) in this ticket.

A new issue would just refer back to #12352 anyway.

**Comment by sdavids** [comment]
> In the React Compiler use-case […] we can definitely […] add docs

At least this should be documented, considering it it the _Future of React_ ™.

**Comment by marcebdev** [comment]
It is what it is, but after 5mo of community discussion here the conclusion seems entirely tone deaf. I again don't mean to come off rude, even though the conclusion you left could be taken as so, but the effective perspective we're left with given such a conclusion is that:

RR decided to use a plugin, this plugin replaces the "official" Vite React plugin, yet does not maintain compatibility with it (aka RR came in an inadvertently broke things). Wether this was an oversight or whatnot is unknown but I'd think I'd have to imagine it is.

To write a conclusion that has no acknowledgment of what happened nor of the last 5mo of discussion, and to shutter the current feedback because it's not perceived as a "difficulty" and therefor not a bug is extremely tone deaf.

I think many people are excited you are trying to become cohesive with a tool like Vite and take a sort of approach as a meta-framework on top of the most popular and liked tools right now, I know I am. However, authoritatively deciding what is/isn't "correct" to you for a community you're joining that has established standard is very much not a way to become cohesive or to address this sort of thing in it's own right.

Maybe you're logical conclusion is even correct that this isn't such a big deal, but this is very very much not how to address nor handle a 5mo old issue by closing it out. Also, what tone does this set? This is uninspiring and I think sets an improper tone for helping build precedent of what developing meta-frameworks on top of Vite could/should look like.

**Comment by pcattori** [comment]
> Isn't it more practical to rename the issue title, if that's the case?
> 
> There are six months' worth of discussion and proposed workarounds (which don't entirely work) in this ticket.
> 
> A new issue would just refer back to [#12352](https://github.com/remix-run/react-router/issues/12352) anyway.

Yea I think you are right. Originally wanted to make sure the original question was the focus, but agree that the discussion here is more valuable to keep in the same issue than to link to in another issue.

**Comment by pcattori** [comment]
To add more context:

I've been collaborating with Arnaud Barré (maintainer of `vite-plugin-react`) since the early days of our Vite plugin on different ways we could interop with `vite-plugin-react` rather than vendoring it. The reason we vendor in `vite-plugin-react` at all is to provide HMR via React Refresh. However, React Router (and Remix) need to add special handling for our framework features, namely route modules. We feel that its our responsibility to provide a great development experience, including HMR, and that its not something you should need to wire up yourself. We've spent many, many hours discussing, designing, and implementing our HMR solution, both within our team and collaborating with Arnaud and the rest of the Vite team. That doesn't mean we necessarily got it right, but we're trying very hard to deliver on that promise.

With that in mind, we'd like React Router (and Remix) to interop well with Babel and especially the React Compiler. So if that's been frustrating, confusing, or broken, its absolutely something we want to fix. Apologies if closing the issue felt dismissive; that was definitely not my intent.

It seems like @pawelblaszczyk5 and @AlemTuzlak  things working via `vite-plugin-babel` (https://github.com/remix-run/react-router/issues/12352#issuecomment-2692706624 and https://github.com/remix-run/react-router/issues/12352#issuecomment-2518455406). I've also been unable to reproduce any errors locally when following that approach.

Is anyone else still experiencing errors/issues with that approach? If so, could you provide a reproduction for us to look at?

**Comment by marcebdev** [comment]
This is super appreciated, I'm glad that it seems in reality everything is going smoothly and it's inspiring that so much thinking took place. I feel like this helps immensely in providing a context on much of the confusion or concern in the discussion leading up to here, and how this arrived to a intended solution.

I think this really stabilizes RR (and the plugin) by showing all the work everyone's done and the effort made to collaborate in the Vite ecosystem, I'm personally excited to see it continue to evolve!

Really appreciate it pcattori, and thank you so much for taking the feedback in good faith so kindly and earnestly!

**Comment by mikkpokk** [comment]
Really appreciate everything! ❤️

> Is anyone else still experiencing errors/issues with that approach? If so, could you provide a reproduction for us to look at?

I’ll do my best to find the cause this week and demonstrate how to reproduce the issue.

**Comment by markdalgleish** [comment]
@mikkpokk Are you able to get back to use with a reproduction?

**Comment by mikkpokk** [comment]
> [@mikkpokk](https://github.com/mikkpokk) Are you able to get back to use with a reproduction?

Sorry, I totally forgot. My RR7 upgrade is still on hold and lives on a separate branch, so it’s been out of sight for a while. There have been a lot of things on my shoulders lately. 

I’ll try to find the time this weekend.

**Comment by marcebdev** [comment]
So far so good for the most part it seems. A note for the `jsxTransform` on the react plugin is that at least in regards to emotion it seems it can be replaced with `@emotion/babel-preset-css-prop` under the babel plugin's presets. Took me a min to figure out what that needed to be replaced with, could be useful to add to documentation on how to transfer that.

Other note, heads up but it doesn't seem like this plugin includes the fix to silence the `Error when using sourcemap for reporting an error: Can't resolve original location of error` warnings, which the react plugin this replaces does. Just an example of disparity

Lastly, it seems to break the banner for `vite-plugin-checker` (it doesn't render), it still seems to run fine so not the biggest deal but something that should likely be fixed as the banner on pages when running dev is nice. Thinking about it maybe it's because how react-router runs the dev server?

I am running the latest of both
`"react-router": "^7.6.0"`
`"vite-plugin-checker": "^0.9.3"`


Current config just for reference:
```
[
      !process.env.VITEST && reactRouter(),
      babel({
        babelConfig: {
          presets: ['@emotion/babel-preset-css-prop'],
          plugins: ['babel-plugin-styled-components', '@emotion/babel-plugin'],
        },
      }),
      tsconfigPaths(),
      checker({
        typescript: { tsconfigPath: './tsconfig.json' },
        eslint: {
          lintCommand: 'eslint "./src/**/*.{js,ts,tsx}" --cache',
          dev: { logLevel: ['warning', 'error'] },
        },
        overlay: { position: 'br', initialIsOpen: false },
        terminal: true,
        enableBuild: true,
      }),
    ]
```
